File: 03/03.py
from pathlib import Path
from typing import List
from functools import reduce
import logging

logging.basicConfig()
log = logging.getLogger(__name__)
log.setLevel(level=logging.WARNING)


def get_priority(elem: str) -> int:
    val = ord(elem) - ord("a") + 1
    return val if val > 0 else ord(elem) - ord("A") + 27


def first_question(input_data: str) -> int:
    lines = input_data.splitlines()
    prio_sum = 0
    for l in lines:
        first_half_set = set(l[: len(l) // 2])
        second_half_set = set(l[len(l) // 2 :])
        common_element = (first_half_set & second_half_set).pop()
        prio_sum += get_priority(common_element)
        log.debug(f"elem: {common_element} prio:{get_priority(common_element)}")
    return prio_sum


def second_question(input_data: str) -> int:
    lines = input_data.splitlines()
    prio_sum = 0
    for n in range(0, len(lines), 3):
        sets = [set(l) for l in lines[n : n + 3]]
        log.debug(sets)
        common_element = reduce(lambda v, a: v & a, sets).pop()
        prio_sum += get_priority(common_element)
        log.debug(f"elem: {common_element} prio:{get_priority(common_element)}")
    return prio_sum


if __name__ == "__main__":

    input_data = (Path(__file__).parent / "INPUT").read_text("UTF8")
    print(first_question(input_data))
    print(second_question(input_data))

File: 03/03.rs
use std::collections::HashSet;
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

fn main() {
    let mut lines = read_lines("./INPUT").expect("error reading file");
    if let Ok(answer1) = first_question(lines) {
        println!("{}", answer1);
    };
    lines = read_lines("./INPUT").expect("error reading file");
    if let Ok(answer2) = second_question(lines) {
        println!("{}", answer2);
    };
}

fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn get_priority(elem: char) -> u32 {
    let ord: u32 = elem.into();
    if ord as i32 - <char as Into<u32>>::into('a') as i32 > 0 {
        return ord - <char as Into<u32>>::into('a') + 1;
    }
    return ord - <char as Into<u32>>::into('A') + 27;
}

fn first_question(lines: io::Lines<io::BufReader<File>>) -> io::Result<u32> {
    let mut prio_sum: u32 = 0;
    for line in lines {
        let l = line?;
        let line_len = l.len();
        let mut first_half_set = HashSet::new();
        first_half_set.extend(l[0..l.len() / 2].chars());

        let mut second_half_set = HashSet::new();
        second_half_set.extend(l[line_len / 2..].chars());
        for c in first_half_set.intersection(&second_half_set) {
            prio_sum += get_priority(*c);
        }
    }
    return Ok(prio_sum);
}

fn second_question(lines: io::Lines<io::BufReader<File>>) -> io::Result<u32> {
    let mut prio_sum: u32 = 0;
    let mut n = 0;
    let mut sets = [HashSet::new(), HashSet::new(), HashSet::new()];
    let sets_len = sets.len();
    for line in lines {
        let l = line?;
        sets[n].extend(l.chars());
        if n == sets_len - 1 {
            let mut accumulator: HashSet<char> = HashSet::new();
            accumulator.extend(sets[0].iter());
            for i in 1..sets_len {
                let mut inter: HashSet<char> = HashSet::new();
                inter.extend(accumulator.intersection(&sets[i]));
                accumulator = inter;
            }
            for c in accumulator.iter() {
                prio_sum += get_priority(*c);
            }
            n = 0;
            sets = [HashSet::new(), HashSet::new(), HashSet::new()];
        } else {
            n += 1;
        }
    }
    return Ok(prio_sum);
}
